### PR TITLE
Pull Request: UNet Segmentation Model Integration for HipMRI Prostate Cancer Study

### PR DESCRIPTION
**Pull request by mgadd02**
This pull request introduces a UNet-based model for semantic segmentation of the prostate gland in MRI images, specifically tailored for the HipMRI Prostate Cancer Study dataset. The model has been simplified to reduce overfitting and achieves a minimum Dice Similarity Coefficient (DSC) of 0.75 on the test set for the prostate label.

Types of Changes
 New feature: Added a new segmentation model and associated scripts.
 Bug fix: N/A
 Breaking change: N/A
 Documentation update: Updated README and added comments for clarity.

Added Files:
- modules.py: For declaring UNET model
- train.py: Training procedure, utilising the modules and dataset files
- predict.py: Uses the trained model and modules.py UNET model for inference
- dataset.py: Organises and proprocesses NIFTI form data for training and prediction

Figure_1

Dice Similarity Coefficient for each class:
Class 0: Dice = 0.9968
Class 1: Dice = 0.9839
Class 2: Dice = 0.9194
Class 3: Dice = 0.9493
Class 4: Dice = 0.8975
Class 5: Dice = 0.8873

### CODE REVIEW

**Comment by yexincheng** [comment]
**This is an initial inspection, no action is required at this point**
1. **Solves problem**: The problem is well solved, all labels reached over 0.88 DICE score. The report is detailed, clear and well-structured. Prediction figures and DICE are reported. The UNet figure in the readme includes a residual unit, but this is not implemented. It would be better to replace this UNet figure with the basic one, aligning with the implementation.
2. **Implementation**: Data augmentation is used. Hpyerparameter is hard-coded rather than using ArgParse. Otherwise, all functions and classes are great.
3. **Design**: Great
4. **Comment**: Very detail and well-structured
5. **Difficulty**: Easy

**Note**
The pull request should be created on topic-recognition branch not main.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Easy. UNet2D (Easy Difficulty)</td><td>-10</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Some/Adequate Meaningful commit messages. </td><td>-1</td></tr>
	<tr><td></td><td>Good Progressive commits. No logical flow.</td><td>-2</td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). Submitted on time.</td><td></td></tr>
	<tr><td></td><td>Feedback action required: Feedback marks possible +2 if the requested changes are made. PR on wrong branch. Correct for merge.</td><td>-2</td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-16</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by gayanku** [comment]
Feedback marks possible +2 if the requested changes are made (see above).

**Comment by shakes76** [comment]
Cant merge because of conflicting changes to main repo files (gitignore). Please update for merge, doesn't affect grade.

**Comment by shakes76** [comment]
Segmentations not one hot encoded -2 additionally, Feedback applied +2

**Comment by mgadd02** [comment]
@shakes76 gitignore removed.

Also I do not understand why outputs must be one hot encoded. It uses class indices. 

Was this specified somewhere ?
It works as intended.

Thanks

**Comment by shakes76** [comment]
Yes it is standard practice to one-hot encode in segmentation and I discussed this in lectures etc. 

Still have conflict in README file of parent recognition directory. Can't merge still.

**Comment by mgadd02** [comment]
@shakes76 I think I fixed the issue with the README.
Thanks. Let me know if anything else I have missed.
